automation/freebsd_repo/FreeBSD_9.2_Repo/src/usr.bin/yacc/yacc.1
.\" Copyright (c) 1989, 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Robert Paul Corbett.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)yacc.1	5.8 (Berkeley) 5/24/93
.\" $FreeBSD: releng/9.2/usr.bin/yacc/yacc.1 216370 2010-12-11 08:32:16Z joel $
.\"	$OpenBSD: yacc.1,v 1.14 2001/05/01 17:58:05 aaron Exp $
.\"
.Dd May 24, 1993
.Dt YACC 1
.Os
.Sh NAME
.Nm yacc
.Nd an LALR(1) parser generator
.Sh SYNOPSIS
.Nm
.Op Fl dlrtvy
.Op Fl b Ar file_prefix
.Op Fl o Ar output_filename
.Op Fl p Ar symbol_prefix
.Ar filename
.Sh DESCRIPTION
The
.Nm
utility reads the grammar specification in the file
.Ar filename
and generates an LR(1) parser for it.
The parsers consist of a set of LALR(1) parsing tables and a driver routine
written in the C programming language.
The
.Nm
utility normally writes the parse tables and the driver routine to the file
.Pa y.tab.c .
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl b Ar file_prefix
Change the prefix prepended to the output file names to
the string denoted by
.Ar file_prefix .
The default prefix is the character
.Pa y .
.It Fl d
Cause the header file
.Pa y.tab.h
to be written.
.It Fl l
If the
.Fl l
option is not specified,
.Nm
will insert #line directives in the generated code.
The #line directives let the C compiler relate errors in the
generated code to the user's original code.
If the
.Fl l
option is specified,
.Nm
will not insert the #line directives.
Any #line directives specified by the user will be retained.
.It Fl o Ar output_filename
Cause
.Nm
to write the generated code to
.Ar output_filename
instead of the default file,
.Pa y.tab.c .
.It Fl p Ar symbol_prefix
Change the prefix prepended to yacc-generated symbols to
the string denoted by
.Ar symbol_prefix .
The default prefix is the string
.Pa yy .
.It Fl r
Cause
.Nm
to produce separate files for code and tables.
The code file
is named
.Pa y.code.c ,
and the tables file is named
.Pa y.tab.c .
.It Fl t
Change the preprocessor directives generated by
.Nm
so that debugging statements will be incorporated in the compiled code.
.It Fl v
Cause a human-readable description of the generated parser to
be written to the file
.Pa y.output .
.It Fl y
NOOP for bison compatibility.
.Nm
is already designed to be POSIX yacc compatible.
.El
.Sh ENVIRONMENT
.Bl -tag -width ".Ev TMPDIR"
.It Ev TMPDIR
Name of directory where temporary files are to be created.
.El
.Sh TABLES
The names of the tables generated by this version of
.Nm
are
.Va yylhs , yylen , yydefred , yydgoto , yysindex ,
.Va yyrindex , yygindex , yytable ,
and
.Va yycheck .
Two additional tables,
.Va yyname
and
.Va yyrule ,
are created if
.Dv YYDEBUG
is defined and non-zero.
.Sh FILES
.Bl -tag -width "Pa /tmp/yacc.aXXXXXXXXXX" -compact
.It Pa y.code.c
.It Pa y.tab.c
.It Pa y.tab.h
.It Pa y.output
.It Pa /tmp/yacc.aXXXXXXXXXX
.It Pa /tmp/yacc.tXXXXXXXXXX
.It Pa /tmp/yacc.uXXXXXXXXXX
.El
.Sh DIAGNOSTICS
If there are rules that are never reduced,
the number of such rules is reported on standard error.
If there are any
.Tn LALR(1)
conflicts,
the number of conflicts is reported on standard error.
.Sh SEE ALSO
.Xr yyfix 1
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.2 .
.Sh HISTORY
A
.Nm
command appeared in PWB UNIX.


automation/freebsd_repo/FreeBSD_9.2_Repo/src/usr.bin/ctags/yacc.c
/*
 * Copyright (c) 1987, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if 0
#ifndef lint
static char sccsid[] = "@(#)yacc.c	8.3 (Berkeley) 4/2/94";
#endif
#endif

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/9.2/usr.bin/ctags/yacc.c 216370 2010-12-11 08:32:16Z joel $");

#include <ctype.h>
#include <limits.h>
#include <stdio.h>

#include "ctags.h"

/*
 * y_entries:
 *	find the yacc tags and put them in.
 */
void
y_entries(void)
{
	int	c;
	char	*sp;
	bool	in_rule;
	char	tok[MAXTOKEN];

	in_rule = NO;

	while (GETC(!=, EOF))
		switch (c) {
		case '\n':
			SETLINE;
			/* FALLTHROUGH */
		case ' ':
		case '\f':
		case '\r':
		case '\t':
			break;
		case '{':
			if (skip_key('}'))
				in_rule = NO;
			break;
		case '\'':
		case '"':
			if (skip_key(c))
				in_rule = NO;
			break;
		case '%':
			if (GETC(==, '%'))
				return;
			(void)ungetc(c, inf);
			break;
		case '/':
			if (GETC(==, '*') || c == '/')
				skip_comment(c);
			else
				(void)ungetc(c, inf);
			break;
		case '|':
		case ';':
			in_rule = NO;
			break;
		default:
			if (in_rule || (!isalpha(c) && c != '.' && c != '_'))
				break;
			sp = tok;
			*sp++ = c;
			while (GETC(!=, EOF) && (intoken(c) || c == '.'))
				*sp++ = c;
			*sp = EOS;
			getline();		/* may change before ':' */
			while (iswhite(c)) {
				if (c == '\n')
					SETLINE;
				if (GETC(==, EOF))
					return;
			}
			if (c == ':') {
				pfnote(tok, lineno);
				in_rule = YES;
			}
			else
				(void)ungetc(c, inf);
		}
}

/*
 * toss_yysec --
 *	throw away lines up to the next "\n%%\n"
 */
void
toss_yysec(void)
{
	int	c;			/* read character */
	int	state;

	/*
	 * state == 0 : waiting
	 * state == 1 : received a newline
	 * state == 2 : received first %
	 * state == 3 : received second %
	 */
	lineftell = ftell(inf);
	for (state = 0; GETC(!=, EOF);)
		switch (c) {
		case '\n':
			++lineno;
			lineftell = ftell(inf);
			if (state == 3)		/* done! */
				return;
			state = 1;		/* start over */
			break;
		case '%':
			if (state)		/* if 1 or 2 */
				++state;	/* goto 3 */
			break;
		default:
			state = 0;		/* reset */
			break;
		}
}
